### ISSUE TITLE
amsmath with options fleqn,leqno does not shift tag for long equation

### ISSUE THREAD

**Issue by gabor-braun** [latex3/latex2e#115]
## Brief outline of the bug

The example below contains some equations with the amsmath package.  Between the equations
vertical lines are added to indicate the page boundary (there are probably much better methods to do this), and some equations are visibly stick out to the right, and the log file reports overfull \hboxes.
Curiously Equation (5) seems to fit in one line despite being longer than Equations (1) and (2):
it has an additional `+9` at the end.

Removing the option `fleqn`, no overfull boxes are reported, and the equation numbers get shifted.

## Minimal example showing the bug

```latex
\RequirePackage{latexbug}
\documentclass{article}
\usepackage[fleqn,leqno]{amsmath}
\begin{document}
\begin{gather}
  0000000000000 + 000000000000000 = 88888888888888888 - 888888888888
  \\
  0000000000000 + 000000000000000 = 88888888888888888 - 888888888888
\end{gather}
\noindent\vrule\strut\hfill\vrule
\begin{align}
  A &= B & C &= D \\
  0000000000000 + 000000000 &= 88888888888888888 - 88888888888
\end{align}

\begin{equation}
  0000000000000 + 000000000000000 = 88888888888888888 - 888888888888
  + 9
\end{equation}
\noindent\vrule\strut\hfill\vrule
\begin{equation}
  \tag{A}
  0000000000000 + 000000000000000 = 88888888888888888 - 888888888888
  + 9
\end{equation}
\end{document}
```

## Log file (required) and possibly PDF file  
PDF for the example above:
[test-amsmath-fleqn.pdf](https://github.com/latex3/latex2e/files/2780312/test-amsmath-fleqn.pdf)

The variant without the option `fleqn`:
[test-amsmath-no-fleqn.pdf](https://github.com/latex3/latex2e/files/2780339/test-amsmath-no-fleqn.pdf)

Log for the example above (with option fleqn)
```log
This is pdfTeX, Version 3.14159265-2.6-1.40.19 (TeX Live 2019/dev/Debian) (preloaded format=pdflatex 2018.12.21)  21 JAN 2019 20:37
entering extended mode
 restricted \write18 enabled.
 file:line:error style messages enabled.
 %&-line parsing enabled.
**\input test-amsmath-fleqn.tex
(./test-amsmath-fleqn.tex
(/usr/share/texlive/texmf-dist/tex/latex/latexbug/latexbug.sty
Package: latexbug 2017/10/12 v1.0d Bug-classification
)
(/usr/share/texlive/texmf-dist/tex/latex/base/article.cls
Document Class: article 2018/09/03 v1.4i Standard LaTeX document class
(/usr/share/texlive/texmf-dist/tex/latex/base/size10.clo
File: size10.clo 2018/09/03 v1.4i Standard LaTeX file (size option)
)
\c@part=\count80
\c@section=\count81
\c@subsection=\count82
\c@subsubsection=\count83
\c@paragraph=\count84
\c@subparagraph=\count85
\c@figure=\count86
\c@table=\count87
\abovecaptionskip=\skip41
\belowcaptionskip=\skip42
\bibindent=\dimen102
)
(/usr/share/texlive/texmf-dist/tex/latex/amsmath/amsmath.sty
Package: amsmath 2018/12/01 v2.17b AMS math features
\@mathmargin=\skip43

For additional information on amsmath, use the `?' option.
(/usr/share/texlive/texmf-dist/tex/latex/amsmath/amstext.sty
Package: amstext 2000/06/29 v2.01 AMS text

(/usr/share/texlive/texmf-dist/tex/latex/amsmath/amsgen.sty
File: amsgen.sty 1999/11/30 v2.0 generic functions
\@emptytoks=\toks14
\ex@=\dimen103
))
(/usr/share/texlive/texmf-dist/tex/latex/amsmath/amsbsy.sty
Package: amsbsy 1999/11/29 v1.2d Bold Symbols
\pmbraise@=\dimen104
)
(/usr/share/texlive/texmf-dist/tex/latex/amsmath/amsopn.sty
Package: amsopn 2016/03/08 v2.02 operator names
)
\inf@bad=\count88
LaTeX Info: Redefining \frac on input line 223.
\uproot@=\count89
\leftroot@=\count90
LaTeX Info: Redefining \overline on input line 385.
\classnum@=\count91
\DOTSCASE@=\count92
LaTeX Info: Redefining \ldots on input line 482.
LaTeX Info: Redefining \dots on input line 485.
LaTeX Info: Redefining \cdots on input line 606.
\Mathstrutbox@=\box27
\strutbox@=\box28
\big@size=\dimen105
LaTeX Font Info:    Redeclaring font encoding OML on input line 729.
LaTeX Font Info:    Redeclaring font encoding OMS on input line 730.
\macc@depth=\count93
\c@MaxMatrixCols=\count94
\dotsspace@=\muskip10
\c@parentequation=\count95
\dspbrk@lvl=\count96
\tag@help=\toks15
\row@=\count97
\column@=\count98
\maxfields@=\count99
\andhelp@=\toks16
\eqnshift@=\dimen106
\alignsep@=\dimen107
\tagshift@=\dimen108
\tagwidth@=\dimen109
\totwidth@=\dimen110
\lineht@=\dimen111
\@envbody=\toks17
\multlinegap=\skip44
\multlinetaggap=\skip45
\mathdisplay@stack=\toks18
LaTeX Info: Redefining \[ on input line 2844.
LaTeX Info: Redefining \] on input line 2845.
)
(./test-amsmath-fleqn.aux)
\openout1 = `test-amsmath-fleqn.aux'.

LaTeX Font Info:    Checking defaults for OML/cmm/m/it on input line 4.
LaTeX Font Info:    ... okay on input line 4.
LaTeX Font Info:    Checking defaults for T1/cmr/m/n on input line 4.
LaTeX Font Info:    ... okay on input line 4.
LaTeX Font Info:    Checking defaults for OT1/cmr/m/n on input line 4.
LaTeX Font Info:    ... okay on input line 4.
LaTeX Font Info:    Checking defaults for OMS/cmsy/m/n on input line 4.
LaTeX Font Info:    ... okay on input line 4.
LaTeX Font Info:    Checking defaults for OMX/cmex/m/n on input line 4.
LaTeX Font Info:    ... okay on input line 4.
LaTeX Font Info:    Checking defaults for U/cmr/m/n on input line 4.
LaTeX Font Info:    ... okay on input line 4.

Overfull \hbox (2.7784pt too wide) detected at line 9
[]
 []


Overfull \hbox (2.7784pt too wide) in alignment at lines 9--9
[] [] 
 []


Overfull \hbox (7.53094pt too wide) detected at line 14
[]
 []

[1

{/var/lib/texmf/fonts/map/pdftex/updmap/pdftex.map}]
(./test-amsmath-fleqn.aux) ) 
Here is how much of TeX's memory you used:
 1314 strings out of 493817
 19296 string characters out of 6156336
 74401 words of memory out of 5000000
 5104 multiletter control sequences out of 15000+600000
 4094 words of font info for 16 fonts, out of 8000000 for 9000
 417 hyphenation exceptions out of 8191
 27i,8n,21p,237b,486s stack positions out of 5000i,500n,10000p,200000b,80000s
</usr/share/texlive/texmf-dist/fonts/type1/public/a
msfonts/cm/cmmi10.pfb></usr/share/texlive/texmf-dist/fonts/type1/public/amsfont
s/cm/cmr10.pfb></usr/share/texlive/texmf-dist/fonts/type1/public/amsfonts/cm/cm
sy10.pfb>
Output written on test-amsmath-fleqn.pdf (1 page, 26320 bytes).
PDF statistics:
 20 PDF objects out of 1000 (max. 8388607)
 13 compressed objects within 1 object stream
 0 named destinations out of 1000 (max. 500000)
 1 words of extra memory for PDF output out of 10000 (max. 10000000)
```

**Comment by RuixiZhang42** [comment]
Not a bug but by design.

The glues in `gather` and `align` are frozen (they no longer have the ability to stretch or shrink). But the glues in `equation` are still flexible.

When you have `fleqn`, the first two environments `gather` and `align` will hang their content from the left margin by `\eqnshift@`, which is a dimension register (not allowed to shrink). Since `\eqnshift@` is large enough for the tags, the tags stay on the same line, and overly long equations stick out with overfull messages. Whilst the last two `equation`s hang their contents by `\@mathmargin`, which is `\leftmargini minus\leftmargini` (allowed to shrink to&nbsp;0), i.e., `equation` will try to squeeze everything and does not honor “math indent” strictly. It succeed with `(5)` but not with `(A)`, because `(A)` was just too wide.

When you don’t have `fleqn`, the first two follow their “centered equation” algorithm and decide to shift the tags because the lines must be at least `\mintagsep` away from the tags. No overfull messages this time because they fit just enough. Also, `equation` now has no `\mathindent` to work with, so it cannot shrink “indent” to squeeze everything. So it decided to typeset `(5)` and `(A)` on the next lines.

IMHO, everything works as expected.

**Comment by gabor-braun** [comment]
Just to repeat: this bug is about inconsistent behavior, which is hard to believe to be either expected or by design.  As far as I understand, the argument given for "not bug by design" is "design = whatever the current code does", which is not convincing.

The inconsistencies are
1) With option `fleqn` in `equation` a too long equation sticks out to the left indentation.  But in `gather` and `align` it sticks out to the text area on the right.  IMHO sticking out to the left indentation is better.
2) With option `fleqn` in `equation` a long equation gets squeezed if necessary to avoid shifting the tag.  (But if the tag is shifted, the equation is squeezed less, compare (5) and (A) in the example, which differ only in the tag.)  However, the equation is not squeezed without `fleqn`.

In summary, this bug is about equations being typeset suboptimal even among the choices, which the current amsmath code could typeset.

**Comment by RuixiZhang42** [comment]
@gabor-braun The point I&nbsp;am trying to make is that the `equation` environment behaves fundamentally differently from `gather`, `align`, aka the `amsmath` environments. Every glue must be frozen inside each field of the `amsmath` environments for measurement and these environments also fix the left indentation in `fleqn` mode. But for `equation`, TeX needs no help from `amsmath` to do the “measurement” and it will make use of any shrinkable glues.

If you add `\mathindent=25pt\relax` after loading `amsmath` in `fleqn,leqno` mode, the (A) equation will now stick out to the right and not stick out to the left indentation. Previously it was set to `25pt minus 25pt`.

I&nbsp;agree this poses inconsistencies but I&nbsp;think the issue is in a different place: The `fleqn` mode prioritizes a constant left indentation, so I&nbsp;would expect `\mathindent=25pt\relax` to make the (5) equation behaving the same as the (A) equation, but it doesn’t. Let equations sticking out to the left indentation in “fleqn” mode doesn’t make much sense to me.
